Add `quantile` and `median`

james77777778 opened the pull request:
Fixes #18409 

Implementing `median` for jax, numpy and torch is straightforward.
For tensorflow backend, this PR adopts and simplifies the `quantile` implementation from TensorFlow Probability to compute `median`.

Addtionally, this PR adds support for multiple axes in torch's `quantile` and `median`.

All the corresponding unit tests have been included.

AakashKumarNain commented on keras/backend/tensorflow/numpy.py:
```diff
@@ -783,6 +788,126 @@ def prod(x, axis=None, keepdims=False, dtype=None):
+        x_ndims = tf.TensorShape(x.shape).rank
+
+        # _make_static_axis_non_negative_list
+        _axis = np.array(axis, dtype="int32")
```
IMO either we should `tfnp` or `tf` for this

james77777778 replied:
Hi @AakashKumarNain 
I have rewritten the logic to use pure python built-ins.

No `import numpy as np` now

fchollet commented on keras/backend/tensorflow/numpy.py:
```diff
@@ -783,6 +788,125 @@ def prod(x, axis=None, keepdims=False, dtype=None):
+    if axis is None:
+        y = tf.reshape(x, [-1])
+    else:
+        x_ndims = tf.TensorShape(x.shape).rank
```
It can generally expected that no tensor can be rank None. So you can just do `len(x.shape)`.

james77777778 replied:
Done

fchollet commented on keras/backend/tensorflow/numpy.py:
```diff
@@ -783,6 +788,125 @@ def prod(x, axis=None, keepdims=False, dtype=None):
+        other_dims = sorted(set(range(x_ndims)).difference(axis))
+        perm = other_dims + list(axis)
+        x_permed = tf.transpose(a=x, perm=perm)
+        if tf.TensorShape(x.shape).is_fully_defined():
```
Likewise this could be `if None not in x.shape` to avoid reliance on TensorShape APIs

james77777778 replied:
Done

fchollet commented on keras/backend/tensorflow/numpy.py:
```diff
@@ -783,6 +788,125 @@ def prod(x, axis=None, keepdims=False, dtype=None):
+        perm = other_dims + list(axis)
+        x_permed = tf.transpose(a=x, perm=perm)
+        if tf.TensorShape(x.shape).is_fully_defined():
+            x_shape = tf.TensorShape(x.shape).as_list()
```
list(x.shape)

james77777778 replied:
Done

fchollet commented on keras/ops/numpy.py:
```diff
@@ -4065,6 +4108,73 @@ def prod(x, axis=None, keepdims=False, dtype=None):
+            quantile. Available methods are `"linear"`, `"lower"`, `"higher"`,
+            `"midpoint"`, and `"nearest"`. Defaults to `"linear"`.
+            If the desired quantile lies between two data points i < j:
+            - `"linear"`: i + (j - i) * fraction, where fraction is the
```
Use backticks around expression (here and on lines below)

james77777778 replied:
Fixed.

fchollet reviewed (commented):
Thank you for the PR!

james77777778 commented:
The failed test should be unrelated.

I have updated the logic of `median` in torch backend to use the native `torch.median`.
Additionally, I have removed the numpy dependency in `keras.backend.torch.numpy` and replaced it with the use of `builtins`.

fchollet reviewed (approved):
The code looks good. Thank you for the great contribution!
